expression/builtin_cast.go

```diff
@@ -1952,14 +1952,8 @@ func WrapWithCastAsDecimal(ctx sessionctx.Context, expr Expression) Expression {
 		if !isnull && err == nil {
 			precision, frac := val.PrecisionAndFrac()
 			castTp := castExpr.GetType()
-			castTp.SetDecimal(frac)
-			castTp.SetFlen(precision)
-			if castTp.GetFlen() > mysql.MaxDecimalWidth {
-				castTp.SetFlen(mysql.MaxDecimalWidth)
-			}
-			if castTp.GetDecimal() > mysql.MaxDecimalScale {
-				castTp.SetDecimal(mysql.MaxDecimalScale)
-			}
+			castTp.SetDecimalUnderLimit(frac)
+			castTp.SetFlenUnderLimit(precision)
```

Comment:
ditto

Reply:
we suppose the setting values are valid after `PrecisionAndFrac`. it would change the logic if they are -1. besides, UpdateFlenAndDecimalUnderLimit aims to maintain valid when update the flen or decimal, where the flen and decimal should be not be -1.